Move CI from GCB to GHA

We want to move all our CI jobs that run on each PR from Google Cloud Build to Github Actions. All GCB pipeline configurations and supporting CI tools are located in https://github.com/gravitational/teleport/tree/master/.cloudbuild.

Unit tests are still failing.

Failures on my linux instance are consistent with GHA. The only difference is that my machine is 8/32 (the same as is used on GCB currently), and the Github machine is 4/16. I've noticed that they never pass initially, but sometimes they pass on the second/third run in the same container. This gives the impression that the reason behind failures is flakiness.

The typical picture is:
[unit-failure-3.json.zip](https://github.com/gravitational/teleport/files/9921833/unit-failure-3.json.zip)

```
===================================================
4929 tests passed. 5 failed, 95 skipped
===================================================
FAIL: github.com/gravitational/teleport/lib/client
FAIL: github.com/gravitational/teleport/lib/client.TestTeleportClient_Login_local/OTP_preferred
FAIL: github.com/gravitational/teleport/lib/client.TestTeleportClient_Login_local
FAIL: github.com/gravitational/teleport/lib/srv/alpnproxy/auth
FAIL: github.com/gravitational/teleport/lib/srv/alpnproxy/auth.TestDialLocalAuthServerNoAvailableServers
===================================================
OUTPUT github.com/gravitational/teleport/lib/client.TestTeleportClient_Login_local
===================================================
=== RUN   TestTeleportClient_Login_local
--- FAIL: TestTeleportClient_Login_local (33.00s)
===================================================
OUTPUT github.com/gravitational/teleport/lib/client.TestTeleportClient_Login_local/OTP_preferred
===================================================
=== RUN   TestTeleportClient_Login_local/OTP_preferred
WARNING: You are using insecure connection to SSH proxy https://localhost:41773
Enter an OTP code from a device:
    api_login_test.go:262:
        	Error Trace:	/workspace/lib/client/api_login_test.go:262
        	Error:      	Received unexpected error:
        	            	Post "https://localhost:41773/v1/webapi/mfa/login/finish": context deadline exceeded
        	Test:       	TestTeleportClient_Login_local/OTP_preferred
    --- FAIL: TestTeleportClient_Login_local/OTP_preferred (5.00s)
===================================================
OUTPUT github.com/gravitational/teleport/lib/srv/alpnproxy/auth.TestDialLocalAuthServerNoAvailableServers
===================================================
=== RUN   TestDialLocalAuthServerNoAvailableServers
    auth_proxy_test.go:38:
        	Error Trace:	/workspace/lib/srv/alpnproxy/auth/auth_proxy_test.go:38
        	Error:      	An error is expected but got nil.
        	Test:       	TestDialLocalAuthServerNoAvailableServers
        	Messages:   	dialLocalAuthServer expected to fail
--- FAIL: TestDialLocalAuthServerNoAvailableServers (0.00s)
===================================================
```

It seems that I've fixed the first one by increasing context timeout, but the second one persists. I have the latest master, just in case.

Sometimes I see this:

```
===================================================
OUTPUT github.com/gravitational/teleport/lib/srv/alpnproxy/auth.TestDialLocalAuthServerNoAvailableServers
===================================================
=== RUN   TestDialLocalAuthServerNoAvailableServers
    auth_proxy_test.go:38:
        	Error Trace:	/workspace/lib/srv/alpnproxy/auth/auth_proxy_test.go:38
        	Error:      	An error is expected but got nil.
        	Test:       	TestDialLocalAuthServerNoAvailableServers
        	Messages:   	dialLocalAuthServer expected to fail
--- FAIL: TestDialLocalAuthServerNoAvailableServers (0.00s)
===================================================
OUTPUT github.com/gravitational/teleport/lib/srv/db.TestAccessRedisAUTHCmd
===================================================
```

[unit-failure-redis.json.zip](https://github.com/gravitational/teleport/files/9921837/unit-failure-redis.json.zip)

--- a/lib/kube/proxy/testing/kube_server/kube_mock.go
+++ b/lib/kube/proxy/testing/kube_server/kube_mock.go
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package proxy
+package kubeserver
 
 import (
 	"bytes"
@@ -32,6 +32,7 @@ import (
 	"github.com/julienschmidt/httprouter"
 	log "github.com/sirupsen/logrus"
 	"golang.org/x/net/http2"
+	v1 "k8s.io/api/authorization/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
@@ -101,7 +102,6 @@ type KubeMockServer struct {
 // More endpoints can be configured
 // TODO(tigrato): add support for other endpoints
 func NewKubeAPIMock() (*KubeMockServer, error) {
-
 	s := &KubeMockServer{
 		router: httprouter.New(),
 		log:    log.NewEntry(log.New()),
@@ -121,6 +121,7 @@ func (s *KubeMockServer) setup() {
 	s.router.UseRawPath = true
 	s.router.POST("/api/:ver/namespaces/:podNamespace/pods/:podName/exec", s.withWriter(s.exec))
 	s.router.GET("/api/:ver/namespaces/:podNamespace/pods/:podName/exec", s.withWriter(s.exec))
+	s.router.POST("/apis/authorization.k8s.io/v1/selfsubjectaccessreviews", s.withWriter(s.selfSubjectAccessReviews))
 	s.server = httptest.NewUnstartedServer(s.router)
 	s.server.EnableHTTP2 = true
 }
@@ -160,7 +161,6 @@ func (s *KubeMockServer) formatResponseError(rw http.ResponseWriter, respErr err
 }
 
 func (s *KubeMockServer) exec(w http.ResponseWriter, req *http.Request, p httprouter.Params) (resp interface{}, err error) {
-
 	q := req.URL.Query()
 
 	request := remoteCommandRequest{
@@ -509,4 +509,21 @@ func v4WriteStatusFunc(stream io.Writer) func(status *apierrors.StatusError) err
 		_, err = stream.Write(bs)
 		return err
 	}
+}
+
+func (s *KubeMockServer) selfSubjectAccessReviews(w http.ResponseWriter, req *http.Request, p httprouter.Params) (resp interface{}, err error) {
+	s1 := &v1.SelfSubjectAccessReview{
+		Spec: v1.SelfSubjectAccessReviewSpec{
+			ResourceAttributes: &v1.ResourceAttributes{
+				Verb: "impersonate",
+			},
+		},
+		Status: v1.SubjectAccessReviewStatus{
+			Allowed: true,
+			Denied:  false,
+			Reason:  "RBAC: allowed",
+		},
+	}
+
+	return s1, nil
 }

--- a/tool/tsh/kube_test.go
+++ b/tool/tsh/kube_test.go
@@ -34,6 +34,7 @@ import (
 	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/lib"
 	"github.com/gravitational/teleport/lib/asciitable"
+	kubeserver "github.com/gravitational/teleport/lib/kube/proxy/testing/kube_server"
 	"github.com/gravitational/teleport/lib/service"
 	"github.com/gravitational/teleport/lib/utils"
 )
@@ -73,6 +74,14 @@ func TestListKube(t *testing.T) {
 			require.NoError(t, err)
 			return len(rootClusters) >= 2
 		}),
+		withNewTeleportOption(
+			// Disables cloud auto-imported labels when running tests in cloud envs such as
+			// Github Actions.
+			// This is required otherwise Teleport will import cloud instance labels and use them
+			// as labels in Kubernetes Service and this test would fail because the output
+			// includes unexpected labels.
+			service.WithIMDSClient(&fakeCloudMetadata{}),
+		),
 	)
 
 	mustLoginSetEnv(t, s)
@@ -152,7 +161,6 @@ func TestListKube(t *testing.T) {
 			require.Contains(t, captureStdout.String(), tc.wantTable())
 		})
 	}
-
 }
 
 func newKubeConfigFile(t *testing.T, clusterNames ...string) string {
@@ -161,7 +169,7 @@ func newKubeConfigFile(t *testing.T, clusterNames ...string) string {
 	kubeConf := clientcmdapi.NewConfig()
 	for _, name := range clusterNames {
 		kubeConf.Clusters[name] = &clientcmdapi.Cluster{
-			Server:                "http://localhost:8080",
+			Server:                newKubeSelfSubjectServer(t),
 			InsecureSkipTLSVerify: true,
 		}
 		kubeConf.AuthInfos[name] = &clientcmdapi.AuthInfo{}
@@ -185,4 +193,39 @@ func formatServiceLabels(labels map[string]string) string {
 
 	sort.Strings(labelSlice)
 	return strings.Join(labelSlice, " ")
+}
+
+type fakeCloudMetadata struct{}
+
+func (f *fakeCloudMetadata) IsAvailable(ctx context.Context) bool {
+	return true
+}
+
+// GetTags gets all of the instance's tags.
+func (f *fakeCloudMetadata) GetTags(ctx context.Context) (map[string]string, error) {
+	return map[string]string{}, nil
+}
+
+// GetHostname gets the hostname set by the cloud instance that Teleport
+// should use, if any.
+func (f *fakeCloudMetadata) GetHostname(ctx context.Context) (string, error) {
+	return "hostname", nil
+}
+
+// GetType gets the cloud instance type.
+func (f *fakeCloudMetadata) GetType() types.InstanceMetadataType {
+	return types.InstanceMetadataTypeDisabled
+}
+
+// GetID gets the cloud instance ID.
+func (f *fakeCloudMetadata) GetID(ctx context.Context) (string, error) {
+	return "id", nil
+}
+
+func newKubeSelfSubjectServer(t *testing.T) string {
+	srv, err := kubeserver.NewKubeAPIMock()
+	require.NoError(t, err)
+	t.Cleanup(func() { srv.Close() })
+
+	return srv.URL
 }

--- a/tool/tsh/tsh_helper_test.go
+++ b/tool/tsh/tsh_helper_test.go
@@ -121,7 +121,7 @@ func (s *suite) setupRootCluster(t *testing.T, options testSuiteOptions) {
 		options.rootConfigFunc(cfg)
 	}
 
-	s.root = runTeleport(t, cfg)
+	s.root = runTeleport(t, cfg, options.newTeleportOptions...)
 	t.Cleanup(func() { require.NoError(t, s.root.Close()) })
 }
 
@@ -187,17 +187,18 @@ func (s *suite) setupLeafCluster(t *testing.T, options testSuiteOptions) {
 	if options.leafConfigFunc != nil {
 		options.leafConfigFunc(cfg)
 	}
-	s.leaf = runTeleport(t, cfg)
+	s.leaf = runTeleport(t, cfg, options.newTeleportOptions...)
 
 	_, err = s.leaf.GetAuthServer().UpsertTrustedCluster(s.leaf.ExitContext(), tc)
 	require.NoError(t, err)
 }
 
 type testSuiteOptions struct {
-	rootConfigFunc func(cfg *service.Config)
-	leafConfigFunc func(cfg *service.Config)
-	leafCluster    bool
-	validationFunc func(*suite) bool
+	rootConfigFunc     func(cfg *service.Config)
+	leafConfigFunc     func(cfg *service.Config)
+	leafCluster        bool
+	validationFunc     func(*suite) bool
+	newTeleportOptions []service.NewTeleportOption
 }
 
 type testSuiteOptionFunc func(o *testSuiteOptions)
@@ -223,6 +224,12 @@ func withLeafCluster() testSuiteOptionFunc {
 func withValidationFunc(f func(*suite) bool) testSuiteOptionFunc {
 	return func(o *testSuiteOptions) {
 		o.validationFunc = f
+	}
+}
+
+func withNewTeleportOption(opt service.NewTeleportOption) testSuiteOptionFunc {
+	return func(o *testSuiteOptions) {
+		o.newTeleportOptions = append(o.newTeleportOptions, opt)
 	}
 }
 
@@ -261,8 +268,8 @@ func newTestSuite(t *testing.T, opts ...testSuiteOptionFunc) *suite {
 	return s
 }
 
-func runTeleport(t *testing.T, cfg *service.Config) *service.TeleportProcess {
-	process, err := service.NewTeleport(cfg)
+func runTeleport(t *testing.T, cfg *service.Config, opts ...service.NewTeleportOption) *service.TeleportProcess {
+	process, err := service.NewTeleport(cfg, opts...)
 	require.NoError(t, err)
 	require.NoError(t, process.Start())
 	t.Cleanup(func() {

